mayastor/src/bdev/nexus/nexus_metadata.rs

```diff
@@ -0,0 +1,620 @@
+
+    #[snafu(display("{}", error))]
+    NexusChildError { error: String },
+}
+
+#[derive(Debug, Deserialize, PartialEq, Default, Serialize, Copy, Clone)]
+pub struct MetaDataHeader {
+    /// Signature identifying this as a MetaDataHeader object
+    pub signature: [u8; 8],
+    /// Size of this header in bytes
+    pub header_size: u32,
+    /// CRC-32 checksum of this header
+    pub self_checksum: u32,
+    /// Current object generation counter
+    pub generation: u64,
+    /// Absolute location (LBA) of this header on disk
+    pub self_lba: u64,
+    /// Offset of start of index table relative to self_lba
+    pub index_start: u64,
+    /// Number of valid entries in the index
+    pub used_entries: u32,
+    /// Maximum number of entries that the index can contain
+    pub max_entries: u32,
+    /// Size of an index entry in bytes
+    pub entry_size: u32,
+    /// CRC-32 checksum of the index table
+    pub index_checksum: u32,
+    /// Offset of first usable data block relative to self_lba
+    pub data_start: u64,
+    /// Offset of last usable data block relative to self_lba
+    pub data_end: u64,
+}
+
+impl MetaDataHeader {
+    pub const METADATA_HEADER_SIZE: u32 = 72;
+    pub const MAX_INDEX_ENTRIES: u32 = 28;
+    pub const INDEX_ENTRY_SIZE: u32 = 36;
+
+    pub fn from_slice(slice: &[u8]) -> Result<MetaDataHeader, MetaDataError> {
+        let mut header: MetaDataHeader =
+            deserialize_from(&mut Cursor::new(slice))
+                .context(DeserializeError {})?;
+
+        if header.header_size != MetaDataHeader::METADATA_HEADER_SIZE {
+            return Err(MetaDataError::HeaderSize {});
+        }
+
+        if header.signature != [0x4d, 0x61, 0x79, 0x61, 0x44, 0x61, 0x74, 0x61]
+        {
+            return Err(MetaDataError::HeaderSignature {});
+        }
+
+        let checksum = header.self_checksum;
+
+        if header.checksum() != checksum {
+            return Err(MetaDataError::HeaderChecksum {});
+        }
+
+        Ok(header)
+    }
+
+    pub fn checksum(&mut self) -> u32 {
+        self.self_checksum = 0;
+        self.self_checksum = crc32::checksum_ieee(&serialize(self).unwrap());
+        self.self_checksum
+    }
+
+    pub fn new(block_size: u32, partition: &GptEntry) -> MetaDataHeader {
+        let index_start = Unsigned::get_aligned_blocks(
+            MetaDataHeader::METADATA_HEADER_SIZE,
+            block_size,
+        );
+
+        let data_start = index_start
+            + Unsigned::get_aligned_blocks(
+                MetaDataHeader::MAX_INDEX_ENTRIES
+                    * MetaDataHeader::INDEX_ENTRY_SIZE,
+                block_size,
+            );
+
+        MetaDataHeader {
+            signature: [0x4d, 0x61, 0x79, 0x61, 0x44, 0x61, 0x74, 0x61],
+            header_size: MetaDataHeader::METADATA_HEADER_SIZE,
+            self_checksum: 0,
+            generation: 0,
+            self_lba: partition.ent_start + 1, /* skip the first block of the
```

Comment:
Lets make sure we keep the terminoligy right -- we are dealing with the metadata itself, dont think that that deals with partitions :)